language/move-model/src/ty.rs

```diff
@@ -505,6 +546,226 @@ impl Type {
     }
 }
 
+/// Holds the unification results between two type instantiations
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct TypeUnification {
+    subst_lhs: BTreeMap<Type, Type>,
+    subst_rhs: BTreeMap<Type, Type>,
+}
+
+impl TypeUnification {
```

Comment:
We already have full type unification as part of ty.rs. It is needed and used for type inference. You may want to check in this draft PR how its used for the invariant problem: https://github.com/diem/diem/pull/8464/files (sorry forgot to share this before I went on vacation).

Maybe one source of confusion in the current type system and how type unification works there is the role of TypeParameter and TypeVar. A TypeParameter is a fixed given type, unification is never working directly on type parameters. Rather unification only works on type variables. A generic function or memory R<X> (X type parameter) is _instantiated_ for a given context by substituting the type parameters with fresh type variables, e.g. X := %0 i.e. R<%0>; subsequently the Substitution::unify function can be called on types containing such variables.

Please check the existing unify implementation which is complete and all we need. Let's discuss in person which use cases you are trying to cover with this new implementation which you think aren't available yet.